From f26098251d469e1d8c29cb3f97127694c3a81e59 Mon Sep 17 00:00:00 2001
From: William Wu <william.wu@rock-chips.com>
Date: Mon, 3 Dec 2018 18:13:15 +0800
Subject: [PATCH] HACK: usb: devio: add delay to fix ss bulk transfer issue

On RK3399/RK3399Pro platforms, We found that when USB 3.0
read/write at the same time in the following test case,
it may easily fail to transfer data with unknown problem.

Host transfer: 24B, 4MB, 4MB, 4MB
Device transfer : 24B, 4MB, 4MB, 4MB

Both Host and Device transfer "24B, 4MB, 4MB, 4M" Repeatedly
until transfer fail.

This patch adds 150us between the 24B and 4MB data to work
around the transfer issue. This patch may affect the USB 3.0
transmission performance, so we mark it as a HACK. We can
revert this patch if we find the root cause and fix this
issue with more reasonable patches.

Change-Id: Iff923ff895396a41a87e6a7df2bfa0072a13e2fc
Signed-off-by: William Wu <william.wu@rock-chips.com>
---
 drivers/usb/core/devio.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/usb/core/devio.c b/drivers/usb/core/devio.c
index 6074ef0acdfe..99bbcfda9872 100644
--- a/drivers/usb/core/devio.c
+++ b/drivers/usb/core/devio.c
@@ -1853,6 +1853,14 @@ static int processcompl(struct async *as, void __user * __user *arg)
 		if (copy_urb_data_to_user(as->userbuffer, urb))
 			goto err_out;
 	}
+
+#ifdef CONFIG_ARCH_ROCKCHIP
+	if (usb_endpoint_xfer_bulk(&urb->ep->desc) &&
+	    urb->actual_length < USB_SG_SIZE &&
+	    as->ps->dev->speed >= USB_SPEED_SUPER)
+		usleep_range(150, 200);
+#endif
+
 	if (put_user(as->status, &userurb->status))
 		goto err_out;
 	if (put_user(urb->actual_length, &userurb->actual_length))
-- 
2.35.3

